include/oneapi/dpl/internal/dynamic_selection.h

```diff
@@ -0,0 +1,179 @@
+    return std::forward<DSPolicy>(dp).invoke_async(e, std::forward<Function>(f), std::forward<Args>(args)...);
+  }
+
+  template<typename DSPolicy, typename Function, typename... Args>
+  inline auto invoke(DSPolicy&& dp, typename DSPolicy::selection_handle_t e, Function&&f, Args&&... args) {
+    return wait_for_all(invoke_async(std::forward<DSPolicy>(dp), e, std::forward<Function>(f), std::forward<Args>(args)...));
+  }
+//ds_policy
+
+  template<typename ScoringPolicy>
+  struct policy {
+    using scoring_policy_t = ScoringPolicy;
+    using native_resource_t = typename scoring_policy_t::native_resource_t;
+    using native_sync_t = typename scoring_policy_t::native_sync_t;
+    using selection_handle_t = typename scoring_policy_t::selection_handle_t;
+    std::shared_ptr<ScoringPolicy> scoring_policy_;
+
+    template<typename... Args>
+    policy(Args&&... args) : scoring_policy_{std::make_shared<scoring_policy_t>(std::forward<Args>(args)...)} {}
+
+    template<typename Property>
+    auto query(const Property &prop) const {
+      return const_cast<const scoring_policy_t &>(*scoring_policy_).query(prop);
+    }
+
+    template<typename Property, typename Other>
+    auto query(const Property &prop, const Other &other) const {
+      return const_cast<const scoring_policy_t &>(*scoring_policy_).query(prop, other);
+    }
+
+    template<typename Property, typename Value>
+    auto report(const Property &prop, const Value &value) const {
+      return const_cast<const scoring_policy_t &>(*scoring_policy_).report(prop, value);
+    }
+
+    template<typename ...Args>
+    auto select(Args&&... args) {
+      return scoring_policy_->select(std::forward<Args>(args)...);
+    }
+
+    template<typename Function, typename ...Args>
+    auto invoke_async(Function&& f, Args&&... args) {
+      return scoring_policy_->invoke_async(std::forward<Function>(f), std::forward<Args>(args)...);
+    }
+
+    template<typename Function, typename ...Args>
+    auto invoke_async(native_resource_t e, Function&& f, Args&&... args) {
+      return scoring_policy_->invoke_async(e, std::forward<Function>(f), std::forward<Args>(args)...);
+    }
+
+    template<typename Function, typename ...Args>
+    auto invoke(Function&& f, Args&&... args) {
+      return scoring_policy_->invoke(std::forward<Function>(f), std::forward<Args>(args)...);
+    }
+
+    template<typename Function, typename ...Args>
+    auto invoke(native_resource_t e, Function&& f, Args&&... args) {
+      return scoring_policy_->invoke(e, std::forward<Function>(f), std::forward<Args>(args)...);
+    }
+
+    auto wait_for_all() {
+      return scoring_policy_->wait_for_all();
+    }
+
+    template<typename SP>
+    friend std::ostream& operator<<(std::ostream &os, const policy<SP>& p);
+  };
+
+  template<typename SP>
+  std::ostream& operator<<(std::ostream &os, const oneapi::dpl::experimental::policy<SP>& q) {
+    os << "DS policy:\n";
+    os << *q.scoring_policy_;
+    return os;
+  }
```

Comment:
Is support for output streams necessary, or is it just a debugging artefact?

If it is removed, do not forget to also remove `#include` directives.

Reply:
Not necessary, especially for beta. For more advanced policies, we may want a way to dump state of policies to inspect later (e.g. useful for something like auto-tuning or even dynamic_load). But not too interesting for current policies. So we can remove.